File: 0410-split-array-largest-sum/0410-split-array-largest-sum.py
class Solution:
    def splitArray(self, nums: List[int], k: int) -> int:
        lo, hi = max(nums), sum(nums)
        ans = -1

        def valid(mid):
            groups = 1
            curSum = 0
            for n in nums:
                if curSum + n > mid:
                    curSum = n
                    groups += 1
                else:
                    curSum += n
            return groups <= k
        
        while lo <= hi:
            mid = (lo + hi) // 2
            if valid(mid):
                ans = mid
                hi = mid - 1
            else:
                lo = mid + 1
        
        return ans

File: 0410-split-array-largest-sum/0410-split-array-largest-sum.cpp
class Solution {
public:
    int splitArray(vector<int>& nums, int k) {
        int lo = 0;
        int hi = 1e9;
        int ans = INT_MAX;

        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;

            if (ok(mid, k, nums)) {
                hi = mid - 1;
                ans = mid;
            } else {
                lo = mid + 1;
            }
        }

        return ans;
    }

private:
    bool ok(int target, int k, vector<int>& nums) {
        int currentSum = 0;
        int count = 1;

        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] > target) return false;
            
            if (currentSum + nums[i] <= target) {
                currentSum += nums[i];
            } else {
                currentSum = nums[i];
                count++;
            }
        }

        return count <= k;
    }
};